lib/rex/proto/ldap/auth.rb

```diff
@@ -0,0 +1,329 @@
+require 'net/ldap'
+require 'net/ldap/dn'
+
+module Rex
+  module Proto
+    module LDAP
+      class Auth
+        NTLM_CONST = Rex::Proto::NTLM::Constants
+        NTLM_CRYPT = Rex::Proto::NTLM::Crypt
+        MESSAGE = Rex::Proto::NTLM::Message
+
+        #
+        # Initialize the required variables
+        #
+        # @param challenge [String] NTLM Server Challenge
+        # @param domain    [String] Domain value used in NTLM
+        # @param server    [String] Server value used in NTLM
+        # @param dnsname   [String] DNS Name value used in NTLM
+        # @param dnsdomain [String] DNS Domain value used in NTLM
+        def initialize(challenge, domain, server, dnsname, dnsdomain)
+          @domain = domain.nil? ? 'DOMAIN' : domain
+          @server = server.nil? ? 'SERVER' : server
+          @dnsname = dnsname.nil? ? 'server' : dnsname
+          @dnsdomain = dnsdomain.nil? ? 'example.com' : dnsdomain
+          @challenge = [challenge.nil? ? Rex::Text.rand_text_alphanumeric(16) : challenge].pack('H*')
+        end
+
+        #
+        # Process the incoming LDAP login requests from clients
+        #
+        # @param user_login [OpenStruct] User login information
+        #
+        # @return auth_info [Hash] Processed authentication information
+        def process_login_request(user_login)
+          auth_info = {}
+
+          if user_login.name.empty? && user_login.authentication.empty? # Anonymous
+            auth_info = handle_anonymous_request(user_login, auth_info)
+          elsif !user_login.name.empty? # Simple
+            auth_info = handle_simple_request(user_login, auth_info)
+          elsif sasl?(user_login)
+            auth_info = handle_sasl_request(user_login, auth_info)
+          else
+            auth_info[:result_code] = Net::LDAP::ResultCodeUnwillingToPerform
```

Comment:
```suggestion
            auth_info = handle_unknown_request(user_login, auth_info)
```
To allow for a future change in behaviour, would it make sense to change this?

Reply:
Per the [RFC](https://datatracker.ietf.org/doc/html/rfc4513) only simple and SASL are valid auth types, any extension is likely to be in `sasl?` and `handle_sasl_request` if/when additional auth is desired.

Reply:
Definitely not a blocker for me

I guess the benefit to the suggestion is it follows the conventions of the other code branches here, and offers the ability for a future class override the base implementation if need be - for whatever reason

Reply:
All good, is a bit YAGNI but that is often the place framework shines. :-)